File: python/btnode.py
class BTNode:
    def __init__(self,elem=None,left=None,right=None,parent=None):
        self.elem = elem
        self.left = left
        self.right = right
        self.parent = parent
    
    def get_elem(self):
        return self.elem
    
    def get_left(self):
        return self.left
    
    def get_right(self):
        return self.right
    
    def get_parent(self):
        return self.parent
    
    def set_elem(self,elem):
        self.elem = elem
    
    def set_left(self,left):
        self.left = left
    
    def set_right(self,right):
        self.right = right
    
    def set_parent(self,parent):
        self.parent = parent
    
    def degree(self):
        degree = 0 
        if left!=None: degree+=1
        if right!=None: degree+=1
        return degree
    
    def is_root(self):
        return self.parent == None
    
    def is_left(self):
        return self.degree() == 0
    
    def is_internal(self):
        return self.degree() == 2
    
    def is_left_child(self):
        return self.parent.get_left() == self
    
    def is_right_child(self):
        return self.parent.get_right() == self
    
    def has_left(self):
        return self.left!=None
    
    def has_right(self):
        return self.right!=None
    
    def get_min(self):
        node = self
        while node.get_left()!=None:
            node = node.get_left()
        return node

File: java/trees/BTNode.java
public class BTNode<E extends Comparable<E>> {
    private E element;
    private BTNode<E> left;
    private BTNode<E> right;
    private BTNode<E> parent;

    public BTNode() {
        this(null,null,null,null);
    }

    public BTNode(E e, BTNode<E> l, BTNode<E> r,BTNode<E> p) {
        element = e;
        left = l;
        right = r;
        parent = p;
    }

    public String element() { return element; }
    public BTNode<E> left() { return left; }
    public BTNode<E> right() { return right; }
    public BTNode<E> parent() { return parent; }

    public void setElement(E s) { this.element = s; }
    public boolean setLeft(BTNode<E> x) { 
        this.left = x;
        if (x!=null) x.setParent(this); 
        return true; 
    }

    public boolean setRight(BTNode<E> x) { 
        this.right = x; 
        if (x!=null) x.setParent(this); 
        return true; 
    }

    public void setParent(BTNode<E> x) { this.parent = x; }

    private int degree() {
        int degree = 0;
        if(left!=null) degree++;
        if(right!=null) degree++;
        return degree;
    }   

    public boolean isRoot() { return parent==null; }
    public boolean isLeaf() { return degree() == 0; }
    public boolean isInternal() { return deree() == 2; }
    public boolean isLeftChild() { return parent.left() == this; }
    public boolean isRightChild() { return parent.right() == this; }
    public boolean hasLeft() { return left!=null; }
    public boolean hasRight() { return right!=null; }

    private BTNode<E> getMin() {
        BTNode<E> x = this;
        while(x.left() !=null) x = x.left();
        return x;
    }

    public String toString() { return element.toString(); }
    
    public int compareTo(E s) {
        return element.compareTo(s);
    }

}